### PR TITLE
CAMEL-20199: camel-core - Remove synchronized blocks

### PR DESCRIPTION
**Pull request by essobedo**
Related to https://issues.apache.org/jira/browse/CAMEL-20199

# Description

For better support of virtual threads, we need to avoid lengthy and frequent pinning by replacing synchronized blocks with ReentrantLocks

# Target

- [ ] I checked that the commit is targeting the correct branch (note that Camel 3 uses `camel-3.x`, whereas Camel 4 uses the `main` branch)

# Tracking
- [ ] If this is a large change, bug fix, or code improvement, I checked there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).

# Apache Camel coding standards and style

- [ ] I checked that each commit in the pull request has a meaningful subject line and body.

- [ ] I have run `mvn clean install -DskipTests` locally and I have committed all auto-generated changes

### CODE REVIEW

**Comment by essobedo** [comment]
Hi, is there anybody that could review this PR? Thank you in advance

**Comment by orpiske** [comment]
Hi @essobedo, I'll take a quick look at it and I'll trigger a full test on my CI. I won't be able to post the results today, though.

**Comment by essobedo** [comment]
> Hi @essobedo, I'll take a quick look at it and I'll trigger a full test on my CI. I won't be able to post the results today, though.

Hi @orpiske, it sounds great thank you

**Review by orpiske — CHANGES_REQUESTED** [review]
I am fine with the approach. As far as I can tell, it's implemented as suggested.

Unfortunately, the way Java implemented VTs causes us to have these excess of Locks (which I kind dislike a bit). 

As such, there's an open question about how much `synchronized` blocks we want to remove. My previous understanding was that this was a problem only on the hot path (however, the code would still be susceptible to thread pinning problems if that code was inadvertently used). 

So, in short: LGTM, but we may still discuss how much `synchronized` we might want to adjust. 

I am only marking as request changes to hear some feedback from @oscerd and @davsclaus

**Comment by essobedo** [comment]
I just wanted to add that if we have to choose between synchronized blocks and locks according to the situation, we need to be very careful during the review of each PR that adds new synchronized blocks as we have to evaluate the locking duration to know if it is an acceptable choice. Having a clear strategy at least allows to ease the review process but there is no bad strategy or good strategy both make sense and have drawbacks so I am waiting for your feedback before starting other PRs for the rest of the code

**Comment by orpiske** [comment]
FYI, I've looked at the test results and they are looking great.

**Comment by essobedo** [comment]
Good to know, thank you @orpiske

**Review by oscerd — APPROVED** [review]

**Comment by oscerd** [comment]
This looks good to me, we just need to apply the same logic everywhere. Thanks.

**Review by davsclaus — APPROVED** [review]

**Review by orpiske — APPROVED** [review]
